### DISCUSSION TITLE
How to "Weave" pixels from different images?

### DISCUSSION THREAD

**Discussion by nocantsin** [ImageMagick/ImageMagick#4968] [Help]
Apologies if this is a trivial question - I've gone through the complete ImageMagick documentation but couldn't find a solution for the following problem:

The individual pixels of four (identically sized) input images should be 'woven' into a new image with twice the horizontal and twice the vertical resolution of the input images.

I.e. if the images had (hypothetically) 2x2 pixels with discrete values:

image 1:
ab
cd

image 2:
ef
gh

image 3:
ij
jk

image 4:
lm
no

...then the resulting image should be:

aebf  
iljm 
cgdh 
knlo

This should work with arbitrary-size matrixes and multi-megapixel files... 

[The actual application would be to re-join Bayer matrix files whose four RGGB color channels had been extracted to separate, lower-resolution TIFF files.]

**Comment by snibgo** [comment]
When asking questions, please always state what version of IM you use, and on what platform. However, that wouldn't change my answer.

> The individual pixels of four (identically sized) input images should be 'woven' into a new image with twice the horizontal and twice the vertical resolution of the input images.

I show how to do that with IM at [Processing Bayer pixels: Combine channel images into Bayer image](http://im.snibgo.com/procbay.htm#combbay). That includes a Windows BAT script rggb2gray.bat, that could easily be converted to bash, if that is your thing.

> This should work with arbitrary-size matrixes and multi-megapixel files...

My script assumes there are four inputs, representing four channels that by default are R, G0, G1 and B. It could be made more general to cope with X\*Y inputs. A possible approach is for the script to build a command with the appropriate processing for each of the X\*Y inputs.

Other material on that page may be of interest to you.

EDIT to add: I use the script on a 12GB laptop with files from a Nikon D800 (approx 7500x5000 pixels), so no problem with multi-megapixel files.

The script includes ...
```
+write msk.png
```
I expect that can be removed.

**Reply by nocantsin** [reply]
Thank you so much! This was exactly the answer I had been looking for. I use IM version 6.9.11 on Linux. Will try to port your batch script to Linux/Unix shell.

**Comment by snibgo** [comment]
For bash, change double "%%" to single "%", change line-continuation caret "^" to backslash "\\", and escape parentheses "\\(" an "\\)".

The Windows BAT construct ...
```
!INNAME:XX=%C0%! 
```
means: take the value of the environment variable INNAME. From that value, change all occurrences of "XX" to the value of the environment variable C0.

**Comment by snibgo** [comment]
An alternative method is (Windows BAT script):
```
magick ^
  pbpx_one_R.miff pbpx_one_G0.miff pbpx_one_G1.miff pbpx_one_B.miff ^
  -size %%[fx:w*2]x%%[fx:h*2] xc: ^
  -colorspace Gray ^
  -insert 0 ^
  -fx "u[1 + i%%2 + 2*(j%%2)].p{floor(i/2),floor(j/2)}" ^
  -alpha off -compress None ^
  -define quantum:format=integer -depth 16 ^
  x.tiff
```
The first four inputs are the half-size channel images. Then we make a full-size image, and insert it at the start of the list (image 0). The channel images are now numbered 1, 2, 3 and 4.

"-fx" sets each pixel for image 0. i and j are the column and row numbers of the pixel. "1 + i%%2 + 2*(j%%2)" is the number of the input image to use: 1, 2, 3 or 4. "floor(i/2)" is the input column, and "floor(j/2)" is the input row.

"-fx" used to be too slow for this, but now it is fast, provided your IM is about v7.1.0-21 or later. It takes about half the time of the script I linked to above.
